### PR TITLE
feat(auto-attach): skip if no codes on aws (CPC-4926)

### PR DESCRIPTION
**Pull request by aciba90**
## Why is this needed?

See CPC0069.

## Please Squash this PR with this commit message

```
feat(auto-attach): skip if no codes on aws (CPC-4926)

On aws, ua-auto-attach.service will only try to auto-attach if the instance identity document[0] contains any billing product or marketplace product code.

Refs:

- [0] https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instance-identity-documents.html#verify-iid
- [1] CPC-4926
- [2] CPC0069

Signed-off-by: Alberto Contreras <alberto.contreras@canonical.com>
```

## Test Steps

```
tox -e behave -- features/ubuntu_pro.feature:191 -D releases=jammy -D machine_types=aws.pro
```

---

- [ ] *(un)check this to re-run the checklist action*

### CODE REVIEW

**Comment by aciba90** [comment]
Force-pushed to fix the integration test that I broke. Other failing integration tests look like unrelated to this change.

**Comment by lucasmoura** [comment]
@aciba90 I thinking I am missing more context around this change. It seems that there is a type of image that we shouldn't auto-attach and this is fixing it. However, which type of image is that and shouldn't our integration test actually verify if auto-attach doesn't work on that image ?

**Comment by aciba90** [comment]
> @aciba90 Shouldn't our integration test actually verify if auto-attach work on the image type defined on the CPC spec ?
> Because by testing it directly on the already used AWS Pro image, we are just guaranteeing that your change doesn't break the process there.
> 
> Additionally, it would be great if we can guarantee that this also doesn't work on generic AWS images

Thanks, @lucasmoura, for pointing that out.

At the moment, we have AWS generic and pro images. The difference is that the latter one comes with `ubuntu-pro-auto-attach` preinstalled. What this PR will allow in the future is to be able to have a single image capable of acting as pro and non-pro. Thus, in the future, both images will contain `ubuntu-pro-auto-attach` preinstalled. Thus, testing current AWS Pro and current AWS Generic + `ubuntu-pro-auto-attach` does cover all cases. I have added a supplemental test covering the later case.

File: features/ubuntu_pro.feature [hunk 1]
```diff
@@ -174,22 +174,84 @@ Feature: Command behaviour when auto-attached in an ubuntu PRO image
...
       | focal   | azure.pro    |
       | focal   | gcp.pro      |
-      | jammy   | aws.pro      |
       | jammy   | azure.pro    |
       | jammy   | gcp.pro      |
-      | noble   | aws.pro      |
       | noble   | azure.pro    |
       | noble   | gcp.pro      |
 
+  Scenario Outline: Auto-attach service works on Pro Machine on aws.pro
```
Comment (a-dubs) [inline features/ubuntu_pro.feature:188]: so this test ensures backwards compatability with existing pro images? and `Scenario Outline: Auto-attach service works on Pro Machine on aws.generic` tests the new functionality right?

File: uaclient/clouds/aws.py [hunk 2]
```diff
@@ -43,10 +42,33 @@ def _get_imds_url_response(self):
...
+            ii_doc = json.loads(
+                self._get_imds_url_response(url, headers=headers)
+            )
+        except json.JSONDecodeError as e:
+            LOG.debug("Error decoding instance identity document: %s", e)
+            return {}
+        return ii_doc
+
+    @property  # type: ignore
+    def is_likely_pro(self) -> bool:
```
Comment (a-dubs) [inline uaclient/clouds/aws.py:65]: :rofl: good function name haha

File: uaclient/clouds/aws.py [hunk 3]
```diff
@@ -43,10 +42,33 @@ def _get_imds_url_response(self):
...
     def identity_doc(self) -> Dict[str, Any]:
-        imds_url_response = self._get_imds_url_response()
+        headers = self._request_imds_v2_token_headers()
+        url = IMDS_URL.format(self._ip_address)
+        imds_url_response = self._get_imds_url_response(url, headers=headers)
         secret_manager.secrets.add_secret(imds_url_response)
         return {"pkcs7": imds_url_response}
 
+    @util.retry(exceptions.CloudMetadataError, retry_sleeps=[0.5, 1, 1])
+    def _get_ii_doc(self) -> Dict:
```
Comment (a-dubs) [inline uaclient/clouds/aws.py:52]: also could you add docstrings to this function and the below `is_likely_pro` function?

**Review by a-dubs — CHANGES_REQUESTED** [review]
left a question, and a small nit for docstrings

File: features/ubuntu_pro.feature (hunk 1 above)
Comment (aciba90) [reply features/ubuntu_pro.feature:188]: `Scenario Outline: Auto-attach service works on Pro Machine on aws.pro` and
`Scenario Outline: Auto-attach service works on Pro Machine on aws.generic` test that pro images and generic images behave correctly with `ubuntu-advantage-pro` installed. 

This service is pre-installed in every Ubuntu Pro image. And at some point in the future, likely, AWS generic images will start shipping `ubuntu-advantage-pro` and not auto-attaching, thanks to this PR.

Comment (aciba90) [reply features/ubuntu_pro.feature:188]: For more context see the attached jira and spec.

File: uaclient/clouds/aws.py (hunk 2 above)
Comment (aciba90) [reply uaclient/clouds/aws.py:65]: Do you have any better suggestion? I think it represents right the feature it holds.

File: uaclient/clouds/aws.py (hunk 3 above)
Comment (aciba90) [reply uaclient/clouds/aws.py:52]: Done.

File: uaclient/clouds/aws.py (hunk 2 above)
Comment (renanrodrigo) [reply uaclient/clouds/aws.py:74]: no this was an actual compliment, keep it

File: features/ubuntu_pro.feature [hunk 4]
```diff
@@ -174,22 +174,84 @@ Feature: Command behaviour when auto-attached in an ubuntu PRO image
...
+      | release |
+      | xenial  |
+      | bionic  |
+      | focal   |
+      | jammy   |
+      | noble   |
+
+  Scenario Outline: Auto-attach service works on Pro Machine on aws.generic
+    Given a `<release>` `<machine_type>` machine with ubuntu-advantage-tools installed
+    When I install ubuntu-advantage-pro
```
Comment (lucasmoura) [inline features/ubuntu_pro.feature:225]: @aciba90 let's instead install the `ubuntu-pro-auto-attach` package. The `ubuntu-advantage-pro` package is now just a transition package to this new one

Comment (aciba90) [reply features/ubuntu_pro.feature]: Done, thanks for the suggestion!

**Review by lucasmoura — APPROVED** [review]

**Review by renanrodrigo — APPROVED** [review]

**Pull request merged**